File: src/manager/radio.cpp
/**
 * SIGMusic Lights 2016
 * Radio class
 */

#include <iostream>
#include <arpa/inet.h>
#include <signal.h>
#include <unistd.h>
#include "radio.h"
#include "../network.h"


using std::cout;
using std::endl;


// Radio pins
#define CE_PIN                  RPI_V2_GPIO_P1_15 // Radio chip enable pin
#define CSN_PIN                 RPI_V2_GPIO_P1_24 // Radio chip select pin

// Cap on the number of light updates per second
#define MAX_FPS         120


RF24 Radio::radio(CE_PIN, CSN_PIN);

struct shared* Radio::s;


Radio::Radio() {

    radio.begin();
    radio.setDataRate(RF24_250KBPS); // 250kbps should be plenty
    radio.setPALevel(RF24_PA_LOW); // Higher power doesn't work on cheap eBay radios
    radio.setRetries(0, 0);
    radio.setAutoAck(false);
    radio.setCRCLength(RF24_CRC_16);
    radio.setPayloadSize(sizeof(packet_t));

    radio.setChannel(RF_CHANNEL);

    radio.openReadingPipe(1, RF_ADDRESS(BASE_STATION_ID));

#ifdef DEBUG
    radio.printDetails();
#endif
}

void ignoreSignal(int signal) {
    return;
}

void Radio::run(struct shared* s) {
    
    Radio::s = s;

    timer_t timer;
    struct sigaction sa;
    struct itimerspec its;

    // Create the timer
    if (timer_create(CLOCK_MONOTONIC, NULL, &timer)) {
        perror("timer_create");
        exit(1);
    }

    // Set up the SIGALRM
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = ignoreSignal;
    sigaction(SIGALRM, &sa, NULL);

    // Calculate the timer interval
    double period = 1.0/MAX_FPS;
    its.it_interval.tv_sec = (time_t) period;
    its.it_interval.tv_nsec = (long) (period * 1000000000) % 1000000000;
    its.it_value.tv_sec = its.it_interval.tv_sec;
    its.it_value.tv_nsec = its.it_interval.tv_nsec;

    // Start the timer
    if (timer_settime(timer, 0, &its, NULL)) {
        perror("timer_settime");
        exit(1);
    }

    sigset_t set;
    sigemptyset(&set);
    sigaddset(&set, SIGALRM);

    while (1) {
        int sig;
        sigwait(&set, &sig);
        puts("Alarm");
        transmitFrame();
    }
}

void Radio::transmitFrame() {

    sem_wait(&s->colors_sem);

    for (int i = 1; i <= 8; i++) {
        
        packet_t msg = {
            {s->colors[i].r, s->colors[i].g, s->colors[i].b}
        };

        radio.openWritingPipe(RF_ADDRESS(i));
        radio.write(&msg, sizeof(msg));
    }

    sem_post(&s->colors_sem);
}

File: software/scheduler/radio.py
"""Define various attributes and functions of the radio network."""

import signal
import threading
import logging
from datetime import datetime

try:
    from RF24 import *
    RF = True
except ImportError:
    RF = False


colors = []
colors_lock = threading.Lock()
colors_lock_lock = threading.Lock()

# The hardcoded RF channel (for now)
RF_CHANNEL = 49

# The endpoint ID of the base station
BASE_STATION_ID = 0

if RF:
    # Radio GPIO pins
    CE_PIN = RPI_V2_GPIO_P1_15  # Chip Enable pin
    CSN_PIN = RPI_V2_GPIO_P1_24  # Chip Select pin

# Cap on the number of light updates per second
# TODO: read this from a config file
MAX_FPS = 1


class RadioNetwork:

    """Manage network operations on the radio.

    Each light is programmed with a unique 1-byte ID. Each has an
    nRF24L01+ radio module configured to listen on an address
    corresponding to its ID: the ASCII characters 'SIGM' followed by
    the one-byte endpoint ID. The Raspberry Pi base station also has an
    nRF24L01+ and follows the same address pattern with an ID of 0.

    The base station transmits an RGB value to each light, in order of
    ID, at a configurable frame rate. These messages are sent
    without acknowledgement for minimal latency.

    """

    num_lights = 8

    def __init__(self):
        """Initialize the network interface."""
        self._logger = logging.getLogger(__name__)
        if RF:
            self._radio = RF24(CE_PIN, CSN_PIN)

            self._radio.begin()
            # 250kbps should be plenty. Lower speed decreases error rate.
            self._radio.setDataRate(RF24_250KBPS)
            # Higher power doesn't work on cheap eBay radios
            self._radio.setPALevel(RF24_PA_LOW)
            self._radio.setRetries(0, 0)
            self._radio.setAutoAck(False)
            self._radio.setCRCLength(RF24_CRC_16)
            self._radio.payloadSize = 3
            self._radio.setChannel(RF_CHANNEL)
            self._radio.openReadingPipe(1, self.rf_address(BASE_STATION_ID))

            if __debug__:
                self._radio.printDetails()

    def start(self):
        """Start periodically transmitting frames."""
        # TODO: use sigwait loop to avoid locking mutex in signal handler
        signal.signal(signal.SIGALRM, self.transmit_frame)
        signal.setitimer(signal.ITIMER_REAL, 1.0/MAX_FPS, 1.0/MAX_FPS)

    def stop(self):
        """Stop periodically transmitting frames."""
        signal.setitimer(signal.ITIMER_REAL, 0)

    def transmit_frame(self, signum, frame):
        """Transmit the color of each light."""
        colors_lock_lock.acquire()
        colors_lock.acquire()
        for i in range(self.num_lights):
            # For now, just print values instead of transmitting
            try:
                rgb = colors[i]
            except IndexError:
                rgb = (0, 0, 0)
            self._logger.debug(rgb)
            if RF:
                msg = bytearray(rgb)
                self._radio.openWritingPipe(rf_address(i))
                self._radio.write(msg)
        colors_lock.release()
        colors_lock_lock.release()

    def rf_address(self, endpoint):
        """Return a 40-bit nRF address."""
        if not 0 <= endpoint < 256:
            raise ValueError
        address_header = "SIGM"
        return int("0x"
                   + "".join(["%0.2x" % ord(i) for i in address_header])
                   + "%0.2x" % endpoint,
                   16)
